### PR TITLE
Bugfix 19441 rulerenderer exp depth

### PR DESCRIPTION
**Pull request by elpaso**
Create a b-tree expression when rule based renderer has more than 50 rules

Fixes #19441 Layers with 80+ rule-based symbology do not render

With tests

Longer explanation: https://issues.qgis.org/issues/19441#note-6

Thanks @rouault for the brainstorming and the help on finding the solution of this issue!

The test for `QgsSQLiteExpressionCompiler` is not really related, but since I wrote it, here it is.

### CODE REVIEW

**Comment by m-kuhn** [comment]
Good catch.

I remember that this kind of complex queries also led to performance degradation on postgres connections (CC @wonder-sk @andreasneumann), so I wonder if we should just turn off server side queries altogether in cases where rule based generated queries grow large like this.

One big question mark I have regarding the proposed approach is if it's ok to insert dataprovider specific optimizations on this level outside the providers. Isn't there a risk, that what's positive for one provider might as well have negative impacts on other ones?

File: tests/src/core/testqgsrulebasedrenderer.cpp [hunk 1]
```diff
@@ -143,6 +143,42 @@ class TestQgsRuleBasedRenderer: public QObject
...
+          rootRule->appendChild( new RRule( nullptr, 0, 0, QStringLiteral( "%1" ).arg( i ) ) );
+        }
+        QgsRuleBasedRenderer r( rootRule );
+        r.startRender( ctx, layer->fields() );
+        r.stopRender( ctx );
+        return r.filter();
+      };
+
+      QCOMPARE( makeFilter( 1 ), QString( "(0)" ) );
+      QCOMPARE( makeFilter( 2 ), QString( "(0) OR (1)" ) );
```
Comment (nyalldawson) [inline tests/src/core/testqgsrulebasedrenderer.cpp]: I don't really understand this - isn't the renderer filter supposed to return a full expression, like "col = 1 or col = 2"?

**Comment by elpaso** [comment]
This test is to check that the ORed rules are organized in a binary tree if they are more than 50. What's in the rule itself is not relevant.

**Comment by nyalldawson** [comment]
Ok, that makes sense!

Looking at the ticket you say "The maximum number of bytes in the text of an SQL statement is limited to SQLITE_MAX_SQL_LENGTH which defaults to 1000000". Should we also add logic in the spatialite provider compiler to fail if the expression exceeds this length?

( I'd argue that most provider compilers should have logic which fails the compilation if the expression contains too many nodes.)

**Comment by elpaso** [comment]
Ideally, this kind of checks should be provider's responsibility.
The problem in this particular issues is that the expression is constructed in the renderer (which is provider agnostic), so we have no way to know whether this b-tree OR join is really necessary or not.

Btw, I think this patch has a negligible impact on performances and will only come into effect in case we have more than 50 rules (I might be wrong but I feel that this is not a very common scenario).
You could probably make it even higher: the problem started to appear at 80 rules, but in that case there were no other filters in action. I'd prefer to leave some room for additional filters and I've chosen 50.

What we should really address (in a separate PR) is better error catching and reporting in case the provider fails in the rendering process: showing a white screen and no messages at all is not very user friendly.

Any objection to merge this patch as it is?

**Comment by m-kuhn** [comment]
Yeah, I think this should really be on a expression compiler level.

@elpaso did you compare performance to completely turning off the expression on the request in this scenario? From what I heard similar scenarios have a (noticeable) negative impact on postgres scenarios as well.

This is simpler and should fix this situation as well (?)

This will also be easier to undo once we have some measures in place like

 * better error reporting
 * simplify expressions (those generated expressions tend to be very naïve)
 * expression compiler level optimization detection

**Comment by nyalldawson** [comment]
I'm ok with seeing this merged as a first step - like @elpaso said, the knowledge of the tree structure of the expression is only known here, so I think it's legitimate to optimise here.

But there's some other followups I'd also like to see, which is to block the spatialite expression compilation when it exceeds the 1000000 size (since that's potentially hit in other circumstances, such as categorized renderer).

And then as @m-kuhn said, I think we should have some logic to abort expression compilation when a certain threshold of complexity is hit.

**Comment by m-kuhn** [comment]
Can you do a short test if this affects performance of postgres in any way (positive or negative)?

**Comment by elpaso** [comment]
@m-kuhn do you mean compare the performances of rendering of a postgis layer with 81 rules without my patch and with my patch?

I imagine that the performances would be greatly affected by the machine(s - because in a real world scenario you won't run posgis on the client) they are executed on, but yes, I think it can be done.

But I also think that it would probably be a waste of time.

I might be wrong here, but I feel that the results greatly depends on the amount of data and the nature of the (80+) rules: if the rules do NOT have an ELSE condition AND the combined OR will limit (and reduce) the total amount of data transfer significantly, then we might have some benefit from filtering on the provider side due to the reduced amount of network I/O, but in case the (80+) rules do not actually filter out any data, then it's just a waste of CPU cycles on both ends (mainly on the server probably).

This is assuming that QGIS must always filter the feature data in the client anyway, because of the rendering rules.

It would be interesting here to dig into the code and check if the QGIS-side feature filtering is executed more times than necessary, I don't know the answer.

**Comment by m-kuhn** [comment]
I think it's useful to run a simple test

* import your test data into postgres
* iterate and measure walltime with the patch
* iterate and measure walltime without the patch
* extra points to do some repeated measures

This, because this kind of expressions have been identified as critical on postgres as well before and because this patch is not isolated to sqlite, so I'd argue that a test will at least give an idea of the side-effects.

Of course much more time could be spent here to optimize and fine-tune, but I don't think I can ask that from you?

**Comment by elpaso** [comment]
@m-kuhn this is all volunteer work, but I'll see what I can do.

**Comment by m-kuhn** [comment]
Thanks @elpaso

**Comment by elpaso** [comment]
I've run the test with and without this patch, on a PG layer and here are the results, the test is a standalone python script which does 10 iterations of map rendering and saving to an 800x800 PNG image, the style contains 79 rules (to make sure it renders without the patch, even if the issue was reported on GPKG and not on PG).

Run without patch: 6.29 secs
Run with patch: 6.23 secs

I did another experiment, and forced-set `filter = QStringLiteral( "TRUE" );` in all cases.

This roughly doubled the time to: 12.24 secs

Possibile explanation: we have 6005 rows in the table but the rules have no ELSE condition and they only match 2307 rows, so in this particular case, the rules filter actually reduces the amount of data that needs to be passed on from the backend to the client.

Does it make sense?

**Comment by m-kuhn** [comment]
Makes a lot of sense, thanks!
I think this confirms the patch has no negative impact on PG.
If you feel like doing one more test, would be nice to see how long it takes with `QgsSettings().setValue('qgis/compileExpressions', False)`

**Comment by elpaso** [comment]
`QgsSettings().setValue('qgis/compileExpressions', False)` scores 15.78 seconds

**Comment by m-kuhn** [comment]
Thanks @elpaso !

**Pull request merged**